compiler/src/dotty/tools/dotc/cc/CheckCaptures.scala

```diff
@@ -593,25 +605,124 @@ class CheckCaptures extends Recheck, SymTransformer:
+          val ares1 = adapt(ares, eres, covariant)
+
+          val resTp =
+            if (ares1 eq ares) && (aargs1 eq aargs) then actual
+            else reconstruct(aargs1, ares1)
+
+          curEnv.captured.asVar.markSolved()
+          (resTp, curEnv.captured)
+        finally
+          curEnv = saved
+
+      /** Adapt type function type `actual` to the expected type.
+       *  @see [[adaptFun]]
+       */
+      def adaptTypeFun(
+          actual: Type, ares: Type, expected: Type,
+          covariant: Boolean, boxed: Boolean,
+          reconstruct: Type => Type): (Type, CaptureSet) =
+        val saved = curEnv
+        curEnv = Env(curEnv.owner, true, CaptureSet.Var(), isBoxed = false, if boxed then null else curEnv)
+
+        try
+          val eres = expected.dealias.stripCapturing match
+            case RefinedType(_, _, rinfo: PolyType) => rinfo.resType
+            case _ => WildcardType
+
+          val ares1 = adapt(ares, eres, covariant)
+
+          val resTp =
+            if ares1 eq ares then actual
+            else reconstruct(ares1)
+
+          curEnv.captured.asVar.markSolved()
+          (resTp, curEnv.captured)
+        finally
+          curEnv = saved
+      end adaptTypeFun
+
+      def adaptInfo(actual: Type, expected: Type, covariant: Boolean): String =
+        val arrow = if covariant then "~~>" else "<~~"
+        i"adapting $actual $arrow $expected"
+
+      /** Destruct a capturing type `tp` to a tuple (cs, tp0, boxed),
+       *  where `tp0` is not a capturing type.
+       * 
+       *  If `tp` is a nested capturing type, the return tuple always represents
+       *  the innermost capturing type. The outer capture annotations can be
+       *  reconstructed with the returned function.
+       */
+      def destructCapturingType(tp: Type, reconstruct: Type => Type = x => x): ((Type, CaptureSet, Boolean), Type => Type) =
+        tp.dealias match
+          case tp @ CapturingType(parent, cs) =>
+            if parent.dealias.isCapturingType then
+              destructCapturingType(parent, res => reconstruct(tp.derivedCapturingType(res, cs)))
+            else
+              ((parent, cs, tp.isBoxed), reconstruct)
+          case actual =>
+            ((actual, CaptureSet(), false), reconstruct)
+
+      def adapt(actual: Type, expected: Type, covariant: Boolean): Type = trace(adaptInfo(actual, expected, covariant), recheckr, show = true) {
+        if expected.isInstanceOf[WildcardType] then actual
+        else
+          val ((parent, cs, actualIsBoxed), recon) = destructCapturingType(actual)
+
+          val needsAdaptation = actualIsBoxed != expected.isBoxedCapturing
+          val insertBox = needsAdaptation && covariant != actualIsBoxed
+
+          val (parent1, cs1) = parent match {
+            case actual @ AppliedType(tycon, args) if defn.isNonRefinedFunction(actual) =>
+              val (parent1, cs1) = adaptFun(parent, args.init, args.last, expected, covariant, insertBox,
```

Comment:
Consider a different name for `cs1` since it's not the same value as the outer `cs1`

Reply:
(same in the instances below)